nifi-nar-bundles/nifi-standard-services/nifi-distributed-cache-services-bundle/nifi-distributed-cache-client-service/src/main/java/org/apache/nifi/distributed/cache/client/CacheClientChannelPoolFactory.java

```diff
@@ -0,0 +1,69 @@
+ */
+package org.apache.nifi.distributed.cache.client;
+
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelOption;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.pool.ChannelPool;
+import io.netty.channel.pool.ChannelPoolHandler;
+import io.netty.channel.pool.SimpleChannelPool;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import org.apache.nifi.components.PropertyValue;
+import org.apache.nifi.controller.ConfigurationContext;
+import org.apache.nifi.event.transport.netty.channel.pool.InitializingChannelPoolHandler;
+import org.apache.nifi.remote.VersionNegotiator;
+import org.apache.nifi.ssl.SSLContextService;
+
+import javax.net.ssl.SSLContext;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Factory for construction of new {@link ChannelPool}, used by distributed cache clients to invoke service
+ * methods.  Cache clients include the NiFi services {@link DistributedSetCacheClientService}
+ * and {@link DistributedMapCacheClientService}.
+ */
+public class CacheClientChannelPoolFactory {
+
+    /**
+     * Instantiate a new netty pool of channels to be used for distributed cache communications
+     *
+     * @param context           the NiFi configuration to be applied to the channel pool
+     * @param versionNegotiator coordinator used to broker the version of the distributed cache protocol with the service
+     * @return a channel pool object from which {@link Channel} objects may be obtained
+     */
+    public static ChannelPool createChannelPool(final ConfigurationContext context, final VersionNegotiator versionNegotiator) {
+        final String hostname = context.getProperty(DistributedSetCacheClientService.HOSTNAME).getValue();
+        final int port = context.getProperty(DistributedSetCacheClientService.PORT).asInteger();
+        final PropertyValue timeoutMillis = context.getProperty(DistributedSetCacheClientService.COMMUNICATIONS_TIMEOUT);
+        final SSLContextService sslContextService = context.getProperty(
+                DistributedSetCacheClientService.SSL_CONTEXT_SERVICE).asControllerService(SSLContextService.class);
+        final SSLContext sslContext = (sslContextService == null) ? null : sslContextService.createContext();
+
+        final EventLoopGroup group = new NioEventLoopGroup();
+        final Bootstrap bootstrap = new Bootstrap();
+        final CacheClientChannelInitializer initializer = new CacheClientChannelInitializer(sslContext, versionNegotiator);
+        bootstrap.group(group)
+                .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, timeoutMillis.asTimePeriod(TimeUnit.MILLISECONDS).intValue())
+                .remoteAddress(hostname, port)
+                .channel(NioSocketChannel.class);
+        final ChannelPoolHandler channelPoolHandler = new InitializingChannelPoolHandler(initializer);
+        return new SimpleChannelPool(bootstrap, channelPoolHandler);
```

Comment:
It would be better to use the `FixedChannelPool`, which supports configuring a maximum number of connections. See `NettyEventSenderFactory` for example usage with additional configuration properties.